core/capi/cjml.cpp

```diff
@@ -0,0 +1,77 @@
+#include <jml.h>
+
+#include <vector>
+
+#include <jml/math/activation_functions.hpp>
+#include <jml/math/vector.hpp>
+
+// This is the global vector of all JML Vector objects.
+std::vector<jml::Vector> vectors;
+
+// Look up a Vector in the array.
+jml::Vector& vector_lookup(unsigned vid) {
+    jml::Vector& v = vectors.at(vid);
+    // TODO -- exception?
+    return v;
+}
+
+int jml_vector_add(unsigned v1, unsigned v2) {
+    jml::Vector& to   = vector_lookup(v1);
+    jml::Vector& from = vector_lookup(v2);
+    to.add(from);
+    return 0;
+}
+
+int jml_vector_add_entry(unsigned vid, int pos, double val) {
+    jml::Vector& v = vector_lookup(vid);
+    v.add_entry(pos, val);
+    return 0;
+}
+
+int jml_vector_apply(unsigned vid, enum JMLActivFunc val) {
+    jml::Vector& v = vector_lookup(vid);
+    jml::ActivationFunction * a;
+    switch (val) {
+        case JACT_FSIG:
+            a = new jml::FastSigmoid();
+        case JACT_SIG:
+            a = new jml::Sigmoid();
+        case JACT_RELU:
+            a = new jml::ReLU();
+    }
+    v.apply(a);
+    delete a;
+    return 0;
+}
+
+double jml_vector_get_entry(unsigned vid, int pos) {
+    jml::Vector& v = vector_lookup(vid);
+    return v.get_entry(pos);
+}
+
+int jml_vector_get_size(unsigned vid) {
+    jml::Vector& v = vector_lookup(vid);
+    return v.get_size();
+}
+
+double jml_vector_set_entry(unsigned vid, int pos, double val) {
+    jml::Vector& v = vector_lookup(vid);
+    v.set_entry(pos, val);
+    return 0;
+}
+
+/**
+ * Returns the newly created vector ID.
+ */
+unsigned jml_vector_init(int length) {
+    unsigned id = vectors.size();
+    vectors.push_back(jml::Vector(length));
+    return id;
+}
+
+int jml_vector_free(unsigned vid) {
+    jml::Vector& v = vector_lookup(vid);
+    // TODO ensure no access can happen in the future
```

Comment:
Probably the most efficient solution (which I think would be good as a separate PR) is to store somewhere that the index is now free, so the next vector to be created through the C API should be stored in the now-free slot.

Reply:
Interesting idea, perhaps use a set that tracks the sparse "holes"? Then, it would be O(log(N)) to check if the Vector referenced by the ID has been destroyed and an unused ID could be obtained by `::begin`

Reply:
Sounds good, I'll open a separate issue for it once this is done.

Reply:
However, this would mean that if a destroyed vector was accidentally rereferenced after a new one has been created in its place, there would be no error. I think this is a non-issue, though (not sure why normal usage would involve deleting and creating vectors).